pennylane/fourier/spectrum.py

```diff
@@ -0,0 +1,272 @@
+
+
+def _get_spectrum(op):
+    r"""Extract the frequencies contributed by an input-encoding gate to the
+    overall Fourier representation of a quantum circuit.
+
+    If :math:`G` is the generator of the input-encoding gate :math:`\exp(-i x G)`,
+    the frequencies are the differences between any two of :math:`G`'s eigenvalues.
+
+    Args:
+        op (~pennylane.operation.Operation): an instance of the `Operation` class
+
+    Returns:
+        list: frequencies contributed by this input-encoding gate
+    """
+    no_generator = False
+    if hasattr(op, "generator"):
+        g, coeff = op.generator
+
+        if isinstance(g, np.ndarray):
+            matrix = g
+        elif hasattr(g, "matrix"):
+            matrix = g.matrix
+        else:
+            no_generator = True
+    else:
+        no_generator = True
+
+    if no_generator:
+        raise ValueError(f"generator of operation {op} is not defined")
+
+    matrix = coeff * matrix
+    # eigenvalues of hermitian ops are guaranteed to be real
+    evals = qml.math.real(np.linalg.eigvals(matrix))
+
+    # compute all differences of eigenvalues
+    frequencies = [np.round(e1 - e2, decimals=8) for e1, e2 in product(evals, evals)]
+    unique_frequencies = list(set(frequencies))
+    return sorted(unique_frequencies)
+
+
+def _join_spectra(spec1, spec2):
+    r"""Join two sets of frequencies that belong to the same input.
+
+    Since :math:`\exp(i a x)\exp(i b x) = \exp(i (a+b) x)`, frequency sets of two gates
+    encoding the same :math:`x` are joined by computing the set of sums of their elements.
+
+    Args:
+        spec1 (list[float]): first spectrum
+        spec2 (list[float]): second spectrum
+    Returns:
+        list[float]: joined spectrum
+    """
+    if not spec1:
+        return sorted(list(set(spec2)))
+    if not spec2:
+        return sorted(list(set(spec1)))
+
+    sums = [s1 + s2 for s1 in spec1 for s2 in spec2]
+    return sorted(list(set(sums)))
+
+
+def spectrum(qnode, encoding_gates=None):
+    r"""Compute the frequency spectrum of the Fourier representation of simple quantum circuits.
+
+    The circuit must only use single-parameter gates of the form :math:`e^{-i x_j G}` as
+    input-encoding gates, which allows the computation of the spectrum by inspecting the gates'
+    generators :math:`G`.
+
+    Gates are marked as input-encoding gates in the quantum function by giving them an ``id``.
+    If two gates have the same ``id``, they are considered
+    to be used to encode the same input :math:`x_j`. The ``encoding_gates`` argument can be used
+    to indicate that only gates with a specific ``id`` should be interpreted as input-encoding gates.
+
+    Args:
+        qnode (pennylane.QNode): a quantum node representing a circuit in which
+            input-encoding gates are marked by their ``id`` attribute
+        encoding_gates (list[str]): list of input-encoding gate ``id`` strings
+            for which to compute the frequency spectra
```

Comment:
Docstring will need to be clearer about what to expect when `encoding_gates` is `None`